### PR TITLE
refactor: command system

### PR DESCRIPTION
**Pull request by Uhuh**
This PR introduces a lot of changes for the command system.

It does...
* Removes the chat commands in favor for the new base command system which introduces 3 new command types
  * Sound command - Users can have sound play whenever this command is ran
  * Choice command - Users can put in options separated by comma's that will get picked randomly and placed into their response text
  * Counter command - Users can have a command increment a value and place that value into a response text.
* Chain commands
  * Users can chain commands together. If I wanted my text command to trigger a sound when it's run, then they can do that. Same for the remaining type of commands.

New view.
image

Chain commands view
image

New command types
image

### CODE REVIEW

**Comment by Uhuh** [comment]
@TODO - Migrate existing users chat commands to text commands.

File: src/app/app.component.ts [hunk 1]
```diff
@@ -213,5 +214,19 @@ export class AppComponent {
...
       VStreamActions.updateState({ partialState: data.value }),
     );
+
+    /**
+     * Whenever the app loads, users sometimes want to reset certain counters.
+     */
+    this.vstreamService.commands$
+      .pipe(
+        first(),
+        map((commands): CounterCommand[] => commands.filter((c): c is CounterCommand => c.type === 'counter' && c.resetOnLaunch)),
```
Comment (skgrush) [inline src/app/app.component.ts]: redundant, given by filter fn

File: src/app/pages/vstream/chat-commands/commands.component.html
Comment (skgrush) [inline src/app/pages/vstream/chat-commands/commands.component.html:1]: Should you have renamed this directory?

File: src/app/pages/vstream/chat-commands/edit-command/chain-command/edit-chain-command/edit-chain-command.component.html [hunk 2]
```diff
@@ -0,0 +1,4 @@
+<div class="chain-command">
+  <app-selector [options]="(options$ | async) ?? []" placeholder="Select a command" [control]="chainCommandControl" />
+  <app-button [style]="'danger'" (click)="deleteChainCommand()">Delete</app-button>
```
Comment (skgrush) [inline src/app/pages/vstream/chat-commands/edit-command/chain-command/edit-chain-command/edit-chain-command.component.html]: `[style]` is a dangerous input name, as it's also used for binding to style attribute

File: src/app/pages/vstream/chat-commands/edit-command/choice-command/choice-command.component.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    payingMembers: new FormControl(false, { nonNullable: true }),
+  });
+
+  constructor() {
+    this.settings.valueChanges
+      .pipe(filter(() => this.settings.valid))
+      .subscribe(settings => {
+        this.vstreamService.updateCommandSettings({
+          ...settings,
+          cooldown: Number(settings.cooldown),
```
Comment (skgrush) [inline src/app/pages/vstream/chat-commands/edit-command/choice-command/choice-command.component.ts]: Why do you need a cast?

Comment (skgrush) [reply src/app/pages/vstream/chat-commands/edit-command/choice-command/choice-command.component.ts]: Ahh are you running into the issue with the dynamically bound `[type]="'number'"` issue in `app-input`? Yeah the only workaround I had for that was to do an `@if(type === 'number') { <input type="number">`

File: src/app/pages/vstream/chat-commands/edit-command/counter-command/counter-command.component.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  });
+
+  readonly permissions = new FormGroup({
+    allUsers: new FormControl(false, { nonNullable: true }),
+    mods: new FormControl(false, { nonNullable: true }),
+    payingMembers: new FormControl(false, { nonNullable: true }),
+  });
+
+  constructor() {
+    this.settings.valueChanges
```
Comment (skgrush) [inline src/app/pages/vstream/chat-commands/edit-command/counter-command/counter-command.component.ts]: Even if you believe these components will live forever I'd _still_ put `takeUntilDestroyed()` on all your subscriptions.

File: src/app/pages/vstream/chat-commands/edit-command/sound-component/sound-command.component.ts [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+      return;
+    }
+
+    const fileURL = await this.toBase64(file);
+
+    if (typeof fileURL !== 'string') {
+      return;
+    }
+
+    this.settings.controls.fileURL.setValue(fileURL);
```
Comment (skgrush) [inline src/app/pages/vstream/chat-commands/edit-command/sound-component/sound-command.component.ts]: Have you considered just storing the `File` ref in the controls?

File: src/app/shared/services/command.interface.ts [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+  payingMembers: boolean;
+};
+
+export interface ChainCommand {
+  id: string;
+  chainCommandID: string | null;
+  chainCommandDelay: number;
+}
+
+interface BaseCommand {
```
Comment (skgrush) [inline src/app/shared/services/command.interface.ts:13]: We might've talked about this, but if this list of commands is exhaustive you can also make a
```ts
export type AnyCommand = ChatCommand | SoundCommand | ChoiceCommand | CounterCommand | VTubeStudioCommand;
```
and this means you can do type discrimination more easily
```ts
function fn(cmd: AnyCommand) {
  if (cmd.type === 'chat') {
    // ts knows cmd is ChatCommand
    // ...
  }
  // ...
}
```

Comment (skgrush) [reply src/app/shared/services/command.interface.ts:13]: doh you're literally already doing this.....

File: src/app/shared/services/command.service.ts [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+      const interval = command.autoRedeemInterval * LOWEST_INTERVAL;
+      const isIntervalValid = interval >= LOWEST_INTERVAL;
+      const autoRedeem = this.autoredeems.get(command.id);
+
+      if (command.autoRedeem && !autoRedeem && isIntervalValid) {
+        this.autoredeems.set(
+          command.id,
+          {
+            interval,
+            timer: setInterval(() => this.handleAutoRedeem(command.id), interval),
```
Comment (skgrush) [inline src/app/shared/services/command.service.ts:87]: If you felt like being more rxjs'y, you could always do `timer: timer(interval).subscribe(() => this.handle...` then when you want to end the timer early just do `redeem.get(id).timer.unsubscribe()`

File: src/app/shared/services/command.service.ts [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+        for (const chainCommand of commands) {
+          const { command, chainCommandDelay } = chainCommand;
+
+          if (!command) {
+            continue;
+          }
+
+          /**
+           * This is a monstrous creation that shouldn't exist
+           * I'm way to lazy to await with promises here
```
Comment (skgrush) [inline src/app/shared/services/command.service.ts]: Move this stuff into a switchMap above, change your forLoop to construct an array of promises, then combineLatest the promise array. If you do this, just make sure to check for an empty array!

File: src/app/shared/state/vstream/vstream.feature.ts [hunk 9]
```diff
@@ -104,22 +113,27 @@ const initialCustomMessage: VStreamCustomMessageState = {
...
-  autoRedeem: false,
-  cooldown: 0,
-  enabled: true,
-  response: 'Change me!',
-};
+/**
+ * Return a default state command for the passed in type.
+ * @param type
+ */
+function changeCommandType(type: CommandTypes) {
```
Comment (skgrush) [inline src/app/shared/state/vstream/vstream.feature.ts:120]: I _always_ want these to be object maps instead but every benchmark I've seen says a switch or if-else is faster 😢

**Review by skgrush — CHANGES_REQUESTED** [review]
I don't know enough on here to provide application review but always happy to be a pedant

**Pull request merged**